src/fe-gtk/maingui.c

```diff
@@ -1084,13 +1084,39 @@ mg_tab_close_cb (GtkWidget *dialog, gint arg1, session *sess)
 	}
 }
 
+typedef struct closed_channel {
+	server *server;
+	char channel[CHANLEN];
+	char key[64];
+} closed_channel;
+
+static GQueue closed_channels = G_QUEUE_INIT;
+
 void
 mg_tab_close (session *sess)
 {
 	GtkWidget *dialog;
 	GSList *list;
 	int i;
 
+	if (hexchat_is_quitting)
+	{
+		closed_channel *chan;
+		while (!g_queue_is_empty(&closed_channels))
```

Comment:
And in any case, [g_queue_free_full](https://developer.gnome.org/glib/stable/glib-Double-ended-Queues.html#g-queue-free-full) exists for this purpose.